### ISSUE TITLE
`delete` TypeScript definition inconsistent with other HTTP method functions

### ISSUE THREAD

**Issue by rhys-vdw** [axios/axios#1365]
#### Summary

Just started migrating to axios. I'm wondering why I can't define a type for the response body [for a `DELETE` request](https://github.com/axios/axios/blob/40b829994c2e407109a38a4cf82703261aa3c22c/index.d.ts#L113). The current typing (introduced in #1061) forces inconsistent syntax:

```ts
const { data: foo } = await axios<Foo>.patch(url, data)

// vs (pick your favourite)

const { data: foo } = await axios.delete(url) as AxiosResponse<Foo>
const { data: foo } = await (axios.delete(url) as AxiosPromise<Foo>)
const { data: foo }: AxiosResponse<Foo> = await axios.delete(url)
const foo = (await axios.delete(url)).data as Foo
```

Some consistency would be good here since both use cases appear to be supported equally.

**Additional thoughts**

Although this syntax is nice and terse (and I'm enjoying using it), it is listed in the ["common mistakes" section of DefinitelyTyped](https://github.com/DefinitelyTyped/DefinitelyTyped#common-mistakes), which I assume are globally applicable rules for typing JS projects.

Specifically:

> If a type parameter does not appear in the types of any parameters, you don't really have a generic function, you just have a disguised type assertion.

The gist of which is that the server can really send back any old object, and there is no way to statically enforce it at TypeScript compile time. So the most "correct" type system to use (given this rule) would be for `data` to always be of type `any`, and force the user to explicitly cast using `as`.

eg.

Here the use of `as` indicates that a type assertion is being made, and that the code cannot be trusted:

```ts
async function getFoo(id: number) {
  const { data } = await axios.get(fooPath(id))
  return data as Foo
}
```

Or if you're really strict about it you can add runtime checks that are backed by the compiler:

```ts
function isFoo(o: any): o is Foo { /* ... */ }

async function getFoo(id: number) {
 const { data } = await axios.get(fooPath(id))
  if (!isFoo(data)) {
    throw new TypeError("That ain't foo!")
  }
  // `data` is now of type `Foo`
  return data
}
```

**cc** @d-fischer

#### Context

- axios version: v0.17.0
- Environment: TypeScript 2.7.1

**Comment by ascott18** [comment]
As a temporary workaround, you can add the following to augment the type definitions

``` ts
declare module "axios" {
  export interface AxiosInstance {
    delete<T = any>(url: string, config?: AxiosRequestConfig): AxiosPromise<T>;
    head<T = any>(url: string, config?: AxiosRequestConfig): AxiosPromise<T>;
  }
}
```

**Comment by nickuraltsev** [comment]
@rhys-vdw Thank you for bringing this up!

**Comment by rhys-vdw** [comment]
Cheers @nickuraltsev. I'm still pretty sure that returning `data: any` rather than a generic return type would be the "most correct" type definition. re: my "additional thoughts" section above. However consistency is the most important thing and the `T= any` type def means that you can still opt into this behaviour.

**Comment by d-fischer** [comment]
`DELETE` is alright, although I don't see how any sensible data would be returned from a `DELETE` request. `HEAD` on the other hand shouldn't ever return any data in the body, i.e. it should always return `AxiosPromise<null>` or similar actually...

**Comment by rhys-vdw** [comment]
Is it unusual to return the deleted resource?

**Comment by ascott18** [comment]
According to the [spec](https://tools.ietf.org/html/rfc7231#section-4.3.5), any 200 response to a DELETE should include a body. If no body is present, the response should be 204.

You are right about HEAD, however - that's my mistake in the PR. I would propose that it be changed to `AxiosPromise<void>` rather than `<null>`

**Comment by rdhelms** [comment]
Has this not been published yet? I'm not seeing the generic type option in `axios@0.18.0`

**Comment by dusan-dragon** [comment]
@rdhelms it is included in axios beta 0.19
